modules/store/src/feature_creator.ts

```diff
@@ -0,0 +1,97 @@
+import { capitalize, isDictionary } from './helpers';
+import { ActionReducer, NonOptional, Primitive } from './models';
+import {
+  createFeatureSelector,
+  createSelector,
+  MemoizedSelector,
+} from './selector';
+
+export type Feature<
+  AppState extends Record<string, any>,
+  FeatureName extends keyof AppState & string,
+  FeatureState extends AppState[FeatureName]
+> = FeatureConfig<FeatureName, FeatureState> &
+  FeatureSelector<AppState, FeatureName, FeatureState> &
+  NestedSelectors<AppState, FeatureState>;
+
+export interface FeatureConfig<FeatureName extends string, FeatureState> {
+  name: FeatureName;
+  reducer: ActionReducer<FeatureState>;
+}
+
+type FeatureSelector<
+  AppState extends Record<string, any>,
+  FeatureName extends keyof AppState & string,
+  FeatureState extends AppState[FeatureName]
+> = {
+  [K in FeatureName as `select${Capitalize<K>}State`]: MemoizedSelector<
+    AppState,
+    FeatureState
+  >;
+};
+
+type NestedSelectors<
+  AppState extends Record<string, any>,
+  FeatureState
+> = FeatureState extends Primitive | unknown[] | Date
+  ? {}
+  : {
+      [K in keyof NonOptional<FeatureState> &
```

Comment:
I think we should create selectors for optional properties if these are assigned a value in the initial state? I would like to hear your thoughts on this though, I can be totally wrong 😅

We could also use `Required` here to oblige that all properties have an initial value/can't be `undefined`.

Reply:
Good point 👍 

If the initial state is defined in the following way:

```ts
interface State {
  foo?: string;
  bar: string;
}

const initialState: State = { /* ... */ };

const feature = createFeature({ name: 'feature', reducer: createReducer(initialState) });
```

From 'types perspective', it cannot be determined whether `foo` is defined in `initialState` or not. 

Now, it seems to me that ignoring the optional fields would confuse many developers 😅 and I more like the second way that you mentioned: to accept only feature states that don't contain optional properties.

In the code, it will look something like this:

```ts
type NotAllowedFeatureStateCheck<
  FeatureState
> = FeatureState extends Required<FeatureState>
  ? unknown
  : 'optional fields are not allowed in feature state';

export function createFeature<...>(
  config: FeatureConfig<...> & NotAllowedFeatureStateCheck<FeatureState>
): Feature<...> { ... } 
```